python/ql/src/experimental/dataflow/internal/Attributes.qll

```diff
@@ -22,7 +23,14 @@ abstract class AttrRef extends Node {
   abstract ExprNode getAttributeNameExpr();
 
   /** Holds if this attribute reference may access an attribute named `attrName`. */
-  predicate mayHaveAttributeName(string attrName) { none() }
+  predicate mayHaveAttributeName(string attrName) {
+    attrName = this.getAttributeName()
+    or
+    exists(Node nodeFrom |
+      localFlow(nodeFrom, this.getAttributeNameExpr()) and
+      attrName = nodeFrom.asExpr().(StrConst).getText()
+    )
+  }
```

Comment:
With the addition of `mayHaveAttributeName`, I'm still not 100% clear on the distinction between that and `getAttributeName` from just reading the QLDoc.

Should the QLDoc be

```ql
  /** Gets the name of the attribute being read or written, if it can be (uniquely) determined statically. */
  abstract string getAttributeName();
```

and that still doesn't _really_ make me happy. With the newest implementation (https://github.com/github/codeql/pull/4423/commits/df447c0af9ec7d62b428ebdfb5d21efcad2c1203), for

```py
getattr(obj, "foo")
```

both `mayHaveAttributeName("foo")` and `"foo" = getAttributeName()` will hold, but for 

```py
key = "foo"
getattr(obj, key)
```
only `mayHaveAttributeName("foo")` will hold. We're capable of (uniquely) determining statically that "foo" will be used as the name of the attribute here, we just don't do it.

---

Ohhh, I see that you provided [context](https://github.com/github/codeql/pull/4423#discussion_r501644537) in the comment, and that this is mostly just copied from JS. Fair enough.

I still think it could be useful to have some QLDoc explaining why both are needed, and when to use which. (I think if you're just reading the QLDoc and member-predicate names, it's hard to figure out what the differences are, and which one to pick).

Reply:
Thanks for the update in https://github.com/github/codeql/pull/4423/commits/b07c7abacc34b71d4426cd12cb55853d565277da @tausbn. How do you feel about the example

```py
key = "foo"
getattr(obj, key)
```

and the fact that `getAttributeName` will not produce any results? :confused: (although it seems reasonable that we should be able to uniquely determine this statically). I'm not against the functionality by any means, just think we need to be clear on what it does.

Reply:
It's gonna be tough, but I think I _may_ be able to sleep at night, even if `getAttributeName` has no results for the above snippet. :wink: 

Do you have an example where this would make a difference for taint tracking? If not, I suggest we leave this as future work.

Reply:
> Do you have an example where this would make a difference for taint tracking? If not, I suggest we leave this as future work.

I'm not arguing we _should_ handle the example above at all :)

> It's gonna be tough, but I think I _may_ be able to sleep at night, even if `getAttributeName` has no results for the above snippet. 

I'm not against the functionality, I just think the current documentation suggests that we would be able to handle the above case, whereas the results will actually be much more limited. What I have in mind is something like

```
  /**
   * Gets the name of the attribute being read or written. Only holds if the attribute name can
   * be determined from examining the expression locally (that is, no references are resolved).
   * For attributes that are computed dynamically, consider using `mayHaveAttributeName` instead.
   */
```

Reply:
> `Only holds if the attribute name can be determined from examining the expression locally (that is, no references are resolved).`

This (wrongly) suggests that `getattr(x,"foo" if 7**2 > 50 else "bar")` would have `bar` as its attribute name (as indeed this is possible to determine statically without resolving references). 

I think we're probably better off _not_ enumerating a set of sufficient conditions under which we _guarantee_ that `getAttributeName` returns a result. Instead, I'll rewrite it to state that this may sometimes fail (for any number of reasons).